lib/concordance_validation/concordance.rb

```diff
@@ -1,34 +1,134 @@
+
+      dedupe_concordance_file
+      populate_database
+
+      # If we have enough memory to keep from thrashing then this may enhance performance.
+      # For now this is too much for desktop/Docker when running a real concordance
+      # so leave it disabled until we can re-evaluate.
+      if TEST_MEMORY_DB
+        mem_db = SQLite3::Database.new ":memory:"
+        # Copy the source DB into the the memory DB
+        Services.logger.info "creating in-memory table..."
+        backup = SQLite3::Backup.new(mem_db, "main", @db, "main")
+        backup.step(-1)
+        backup.finish
+        @db.close
+        @db = mem_db
+      end
+    end
+
+    def populate_database
+      count = @db.execute("SELECT COUNT(*) FROM concordance")[0][0]
+      if count.zero?
+        Services.logger.info "importing data from #{@deduped_concordance_file}..."
+        sqlite3_command = <<~END
+          sqlite3 "#{@db_file}" << EOF
+          .mode tabs
+          .import "#{@deduped_concordance_file}" "concordance"
+          EOF
+        END
+        system(sqlite3_command, exception: true)
+        Services.logger.info "creating indexes..."
+        @db.execute CREATE_VARIANT_INDEX_SQL
+        @db.execute CREATE_CANONICAL_INDEX_SQL
       end
-      @variant_to_canonical.default_proc = ->(_, _) {}
-      @canonical_to_variant.default_proc = ->(_, _) {}
+    end
+
+    def dedupe_concordance_file
+      @deduped_concordance_file = File.join(
+        @db_dir,
+        # Strip off possible double .txt.gz suffix on path to the compressed file.
+        File.basename(@concordance_file).split(".")[0] + "_dedupe.txt"
+      )
+      if File.exist? @deduped_concordance_file
+        Services.logger.info "deduped concordance #{@deduped_concordance_file} already exists, skipping"
+      else
+        cmd = if /\.gz$/.match?(@concordance_file)
+          "zcat -cf #{@concordance_file} | sed '#{SED_DEDUPE_REGEX}' > #{@deduped_concordance_file}"
+        else
+          "sed '#{SED_DEDUPE_REGEX}' #{@concordance_file} > #{@deduped_concordance_file}"
+        end
+        Services.logger.info "deduping #{@concordance_file} to #{@deduped_concordance_file}..."
+        system(cmd, exception: true)
+      end
+    end
+
+    # WARNING: this is for testing only
+    def to_h
+      Hash.new { |h, k| h[k] = [] }.tap do |hash|
+        @db.execute("SELECT variant,canonical FROM concordance").each do |row|
+          hash[row[0]] << row[1]
+        end
+      end
+    end
+
+    def variant_to_canonical(variant)
+      @db.execute("SELECT canonical FROM concordance WHERE variant=? AND canonical!=?", [variant, variant])
+        .flatten
+        .uniq
+    end
+
+    def canonical_to_variant(canonical)
+      @db.execute("SELECT variant FROM concordance WHERE canonical=? AND variant!=?", [canonical, canonical])
+        .flatten
+        .uniq
     end
 
     def file_handler
-      if /\.gz$/.match?(@infile)
+      if /\.gz$/.match?(@concordance_file)
```

Comment:
`@concordance_file.end_with?(".gz")` might be half a cpu-cycle shorter and more Ruby-idiomatic.

Besides, is `file_handler` even used anywhere?

Reply:
No, it's not used. I got rid of it, yay!